keyedcache/api.go

```diff
@@ -0,0 +1,153 @@
+// marked dirty.
+//
+type CachedObject interface {
+	Hold(cntxt *ReqContext)
+	Release(cntxt *ReqContext)
+	MarkDirty(cntxt *ReqContext)
+	IsDirty(cntxt *ReqContext) bool
+	Flush(cntxt *ReqContext) (err error)
+}
+
+// Cache caches objects identified by keys, upto a specified maximum amount of
+// memory.  Objects in the cache are accessed by calling Lookup() with the
+// desired key and Lookup() returns a pointer to an object with a hold on it.
+// When the caller is done with the object it must release the hold.
+//
+// If the object is not currently cached, Cache will create a new object
+// corresponding to the key or reuse an existing one.  It is initialized by a
+// call to Object.Read() before return.  Cache locks the object duing the call
+// to Object.Read() so only one object with a given key is present in the cache
+// and no thread can access it until it is initialized.
+//
+// There is no provision for otherwise locking the object.  In particular, an
+// object can be marked dirty or flushed while multiple threads have a hold on
+// the object.
+//
+// A Cache provides the following methods:
+//
+//   o Lookup() -- lookup an object by key, creating it if necessary; may return
+//     an error if Read() fails; blocks if the cache is full
+//   o Flush() -- flush all objects matched by the function predicate(); return
+//     an error if any of the Object.Write() fails
+//   o Inval() -- invalidate all  objects matched by the function predicate();
+//     InvalFlush determines whether dirty objects are flushed before they are
+//     invalidated; Inval() will block until a held object is released; Inval()
+//     will return an error if flush failed
+//
+type Cache interface {
+	Lookup(cntxt *ReqContext, key Key) (err error)
+	Flush(cntxt *ReqContext, predicate func(cntxt *ReqContext, key *Key) bool) (err error)
+	Inval(cntxt *ReqContext, predicate func(cntxt *ReqContext, key *Key) bool,
+		flushIfDirty InvalFlush) (err error)
+	Free(cntxt *ReqContext)
+}
+
+type InvalFlush int
+
+const (
+	INVAL_DIRTY InvalFlush = iota
+	FLUSH_DIRTY InvalFlush = iota
+)
+
+// Create a new keyed cache whiche consumes a maximum of cacheSz bytes of memory
+// (not including hash chains and some other data).  The caller supplies the
+// size of keys and objects so we can compute the maximum number of objects to
+// cache.
+//
+// The caller also supplies newObject(), which is called by the cache when it
+// wants to create a new object.  Objects must embed a pointer to a CachedObject
+// and initialize that pointer with the value passed in.
+//
+// The *ReqContext is passed through without using it, however the cntxt passed
+// to newObject() will be the cntxt passed to the call to Lookup() that
+// triggered object creation.
+//
+func NewCache(cntxt *ReqContext,
+	newObject func(cntxt *ReqContext, cachedObj *CachedObject) (object Object),
+	cacheSz uint64, objectSz uint64, keySz uint64) (cache Cache)
```

Comment:
What are `objectSz` and `keySz`? How are they related to `Object.SizeOf` and `Key.SizeOf`?